### PR TITLE
Various generative model for OASIS dataset

### PR DESCRIPTION
**Pull request by KasperChenGH**
Implementation of various generative model of OASIS dataset. (VQVAE, Stable Diffusion, Visual Transformer)

This pull request contains the following files: 
`model_ckpt/`: models' checkpoints.
`collected_latents/`: generate latent data for the second stage training.
`log/`: training log. (Can use logger.py to visualize it.)
`visualize/`: visualization of images when training.
`playground/*`: some code to figure out some algorithms.
`README.md / report.pdf`: The file you current see.
`tech_note.md / tech_node.pdf`: The main algorithm I used.

`util.py`: useful function, such as ssim, positional encoding.
`modules.py`: useful network module, such as resblock.
`dataset.py`: get dataset from given folder.
`logger.py`: script that can check the training log.

`model_VAE.py`: VAE and VQVAE model for first stage.
`model_discriminator.py`: The GAN part when training VAE or VQVAE.
`pre_train.py`: Training Script for first stage.

`latent_dataset.py`: Collect the latent data given an first stage mode.
`train.py`: Implement stable diffusion when random generation from VQVAE or VAE model.
`train_ViT.py`: Implement the Visual Transformer (ViT) for random image generation from VQVAE mode.
`model_diffusion.py`: The model of stable diffusion. It's U-Net by the way.

`predict.py`: Reproduce SSIM score on task 8 and generate evolution gif on task 9.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Usage: Good
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- All generative models look good

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves multiple problems! +1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
